ProjFS.Mac/PrjFSKextTests/ShouldHandleFileOpTests.mm

```diff
@@ -0,0 +1,298 @@
+            nullptr, // path
+            KAUTH_FILEOP_RENAME,
+            true, // isDirectory,
+            &testRootHandle,
+            &pid));
+
+    PrjFSProviderUserClient userClient;
+    testRootHandle = InsertVirtualizationRoot_Locked(
+        &userClient,
+        0,
+        self->repoRootVnode.get(),
+        self->repoRootVnode->GetVid(),
+        FsidInode{ self->repoRootVnode->GetMountPoint()->GetFsid(), self->repoRootVnode->GetInode() },
+        self->repoPath.c_str());
+    XCTAssertTrue(VirtualizationRoot_IsValidRootHandle(testRootHandle));
+
+    // With Valid Root Handle we should pass
+    XCTAssertTrue(
+        ShouldHandleFileOpEvent(
+            &self->perfTracer,
+            self->context,
+            self->repoRootVnode.get(),
+            nullptr, // path
+            KAUTH_FILEOP_RENAME,
+            true, // isDirectory,
+            &testRootHandle,
+            &pid));
+}
+
+- (void)testUnsupportedFileSystem {
+    shared_ptr<vnode> testVnode = vnode::Create(self->testMount, "/none");
+    int pid;
+    VirtualizationRootHandle testRootHandle;
+    
+    // Invalid File System should fail
+    XCTAssertFalse(
+        ShouldHandleFileOpEvent(
+            &self->perfTracer,
+            self->context,
+            testVnode.get(),
+            nullptr, // path
+            KAUTH_FILEOP_RENAME,
+            true, // isDirectory,
+            &testRootHandle,
+            &pid));
+}
+
+- (void)testUnsupportedVnodeType {
+    shared_ptr<vnode> testVnodeUnsupportedType = vnode::Create(self->testMount, "/foo", VNON);
+
+    PrjFSProviderUserClient userClient;
+    VirtualizationRootHandle testRootHandle = InsertVirtualizationRoot_Locked(
+        &userClient,
+        0,
+        self->repoRootVnode.get(),
+        self->repoRootVnode->GetVid(),
+        FsidInode{ self->repoRootVnode->GetMountPoint()->GetFsid(), self->repoRootVnode->GetInode() },
+        self->repoPath.c_str());
+    XCTAssertTrue(VirtualizationRoot_IsValidRootHandle(testRootHandle));
+
+    int pid;
+    // Invalid Vnode Type should fail
+    XCTAssertFalse(
+        ShouldHandleFileOpEvent(
+            &self->perfTracer,
+            self->context,
+            testVnodeUnsupportedType.get(),
+            nullptr, // path
+            KAUTH_FILEOP_RENAME,
+            true, // isDirectory,
+            &testRootHandle,
+            &pid));
+}
+
+- (void)testProviderOffline {
+    PrjFSProviderUserClient userClient;
+    VirtualizationRootHandle testRootHandle = InsertVirtualizationRoot_Locked(
+        &userClient,
+        0,
+        self->repoRootVnode.get(),
+        self->repoRootVnode->GetVid(),
+        FsidInode{ self->repoRootVnode->GetMountPoint()->GetFsid(), self->repoRootVnode->GetInode() },
+        self->repoPath.c_str());
+    XCTAssertTrue(VirtualizationRoot_IsValidRootHandle(testRootHandle));
+
+    int pid;
+    // Fail when the provider is not online
+    s_virtualizationRoots[0].providerUserClient = nullptr;
```

Comment:
This should really be an `ActiveProvider_Disconnect()` call. (In which case there needs to be a `vnode_get(self->repoRootVnode.get())` after the insertion above to balance the `vnode_put` in Disconnect; this is a bit ugly and I believe I fixed the insertion to perform the `_get` in #500. I think I'll extract the simpler bug fixes from that PR and resubmit them as a standalone one.)

Reply:
I see that you made this change in #1200.  

I will leave this as-is here and wait to merge this PR in until after you've merged #1200.

Reply:
Actually I see that this is a slightly different suggestion, fixing now